Lilypad-Tech/lilypad
# build: Add release pipeline

### Review Type Requested (choose one):


### Summary

This pull request adds a release workflow that:

- [x] Automatically generates release pull requests and merges them immediately
- [x] Automatically creates releases when the release pull request it merged
- [x] Extends generated release notes with custom notes
- [x] Builds binaries and uploads them to the release

It also makes a few associated changes:

- [x] Update `testnet_deploy_services.yml` to deploy only on releases
- [x] ~~Update `release_docker.yml` to publish only on releases~~. Removed. We also want to publish on push to `main`.
- [x] Update resource provider updater script to expect `-gpu` suffix in release name (and rename script to `lilypad-updater-gpu.sh`)
- [x] Remove CirclCI build and existing GH actions build and release workflows
- [x] Naming clean up to use Go conventions (https://go.dev/doc/effective_go#mixed-caps)

### Task/Issue reference

Closes: #91, closes #168

### Details

#### Goals

Our goals for the release pipeline:

- Release features and fixes on merge to `main`
- Support trunk-based development
- Automate everything as much as possible 

#### release-please

This PR uses [`release-please`](https://github.com/googleapis/release-please-action) to generate release pull requests and create releases when new features or fixes are introduced. `release-please` checks for conventional commit prefixes to determine a semantic version and to decide when to create a release. From their documentation:

> The most important prefixes you should have in mind are:
>
> - `fix:` which represents bug fixes, and correlates to a SemVer patch.
> - `feat:` which represents a new feature, and correlates to a SemVer minor.
> - `feat!:`, or `fix!:`, `refactor!:`, etc., which represent a breaking change (indicated by the `!`) and will result in a SemVer major.

`release-please` does not create release PRs or releases for tasks without user facing (prefixes like `chore:`, `build:`, etc).

#### Automate it more please

A typical `release-please` setup creates a release pull request that a human approves and merges. Once the PR is merged, `release-please` runs a second time and creates a release when it observes its release commit.

We take this a step further by automatically merging the release pull request. After we merge, `release-please` runs a second time to create the release.

Here is a diagram showing an example flow on commits coming into `main` and how the releases workflow responds to each. The diagram assumes we have started at version `1.0.0`:

<img width="1747" alt="release-pipeline" src="https://github.com/Lilypad-Tech/lilypad/assets/7957636/e67e0051-5c98-48e8-9799-657dace28cfd">

We have created a `lilypad-releases` bot (implemented as a GitHub app) to run `release-please` actions. The custom bot is needed to trigger further actions, such as running the workflow a second time on merge. 

In addition, our branch protection rules for `main` require one PR approval and will require passing tests in the near future. The `lilypad-releases` bot cannot approve it's own pull request, so we have the default `github-actions` bot step in for approval.

#### Customization

`release-please` automatically maintains a CHANGELOG, generates pull request text, and generates release notes.

We can customize the pull request text by updating [release-please-config.json](https://github.com/Lilypad-Tech/lilypad/blob/main/release-please-config.json).

We extend the generated release notes with an `extend-notes` job. This job reads a [`release_notes.md`](https://github.com/Lilypad-Tech/lilypad/blob/a0f6a3e21f5f9bf16549addff4ddd91124f74a23/.github/releases/release_notes.md) file and appends it to the generated release notes. We can customize the release notes by updating this file, but note that the changes will persist in the next release unless they are changed again.

### How to test this code?

This pull request is challenging to test in this repository because we need to merge it to see it in action.

I have created a [Lilypad-Tech/test-gh-action-releases](https://github.com/Lilypad-Tech/test-gh-action-releases) repo to demonstrate the release workflow with a few test PRs ready to approve and merge: https://github.com/Lilypad-Tech/test-gh-action-releases/pulls

### Anything else?

Our internal version to capture the design can be found here: https://www.notion.so/lilypadnetwork/Release-Pipeline-70ecc09bed8b406ba33e7e1e4c4e93b3
